keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1045 @@
++	// source in its spec.
++	//
++	// Kubelet will begin trying to rotate the certificate after 50% of the leaf
++	// certificate's lifetime has elapsed, or 24 hours, whichever is shorter.
++	//
++	// The credential bundle is a single file in PEM format.  The first PEM
++	// entry is the private key, and the remaining PEM entries are the
++	// certificate chain issued by the signer (typically, signers will return
++	// their certificate chain in leaf-to-root order).
++	//
++	// The named signer controls chooses the format of the certificate it
++	// issues; consult the signer implementation's documentation to learn how to
++	// use the certificates it issues.
++	//
++	// +featureGate=PodCertificateProjection
++ // +optional
++	PodCertificate *PodCertificateProjection `json:"podCertificate,omitempty" protobuf:"bytes,6,opt,name=podCertificate"`
+}
+```
+
+#### Example certificate bundle written to the pod filesystem
+
+```
+-----BEGIN PRIVATE KEY-----
+<Omitted, so that automation stops mailing the KEP authors to complain>
+-----END PRIVATE KEY-----
+-----BEGIN CERTIFICATE-----
+MIIDSDCCAjCgAwIBAgIQRleU9YIlOqvluwFKzauVoTANBgkqhkiG9w0BAQsFADAV
+MRMwEQYDVQQDEwprdWJlcm5ldGVzMB4XDTIzMTAyOTA1NDU1NVoXDTIzMTAzMDA1
+NTAyMlowADCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAMy7zINPgkDc
+kCrb2EEuxvt6PLjthjnlq9uxl79dhQGl7EgIEgA4Kc+WKZ6uJsYJlJazhuc6egDY
+7nr8s+MATtPV8hV6Rgd/8bMB2ktq1RVhEJ4kRAodSLDLnD4e0IhIZpXSK5DBCXF6
+dRotiCEIGpCQmdMTDYZez6zTfN3E1yuhyw7bSKc2OJj6+MYly6oQwE0T7WVDNqoa
+Ev2ZdHZjePkEhXvGRvm1fyxK4h0UJsaT38/COUW/Dal7Fx6VcJeY2QHI28W+NYP3
+AOdx/7Hc6bbPcYZI/c1HPolUhoX6huiltGMn3yOLsaGjajwXUoE2aKsy3AHxeIkQ
+4UlKQSabpb8CAwEAAaOBqDCBpTAOBgNVHQ8BAf8EBAMCA6gwEwYDVR0lBAwwCgYI
+KwYBBQUHAwIwHwYDVR0jBBgwFoAUuQ7U2/jMkK4nY5R2rPvjyCiC2SQwXQYJKwYB
+BAGDwlMBBFAwTgwHZGVmYXVsdAwHZGVmYXVsdAwHcGVtLXBvZAwkMjc3YjljNDkt
+MzljMy00ZjIzLWIwYjEtZTg4OWVmOTBjMGIyDAtraW5kLXdvcmtlcjANBgkqhkiG
+9w0BAQsFAAOCAQEAQd1G2CHdGPv9Lu9/SD7HPNXwI0K4JrXsWxJQ7Q16IVAuuWIY
+mOdumO9L5PtTrV1widNm99oMEtztQHZoHHIkkypjfW+BB8XwO5er+FpcUCuvZ3w/
+g9Zk9cZUqCdWER0UDl4RCLcS+M6QSq0nUg+JTZKRGJ5fjns3vZHA9Nswp4XRvjt8
+yf6RY6oEJoa10QxB7YiNVmVkqLifhkQiF0Bmz7tONbsjceZ030Cs/wjdssnNzRGL
+UdFfcDKw4QbMGqntJXPFqu86lk0IJwir+aTk/31yDZX1ZyOK0Y/SI7ZVVDeETTA9
+k68gmm9HQCdRI3stW7TC3lB0Cd1XSSMUISIP0g==
+-----END CERTIFICATE-----
+```
+
+### API Server PodIdentity Extension Support
+
+The kube-apiserver is extended to authenticate TLS client certificates that:
+
+* Are signed by the existing `--client-ca-file` CA,
+* Contain empty Name and SubjectAlternateName data, and
+* Contain an X.509 PodIdentity extension.
+
+These certificates are translated to an authenticator.Response that has:
+* `User.Name` set to
+  `system:serviceaccount:{PodIdentity.Namespace}:{PodIdentity.ServiceAccountName}`.
+* `User.Extra["authentication.kubernetes.io/pod-name"]` set to
+  `{PodIdentity.PodName}`
+* `User.Extra["authentication.kubernetes.io/pod-uid"]` set to
+  `{PodIdentity.PodUID}`
+* `User.Extra["authentication.kubernetes.io/node-name"]` set to
```

Comment:
We should also set the node UID here?

Reply:
And also the service account UID? Or I wonder if that'll be used in User.UID instead?

Reply:
Ditto the conversatin about SA / Node UIDs above.